Clang fails to compile the vectorized backends

I have tried using Clang 6.0.1. This is the build script I use:

```
mkdir build
cd build
cmake .. -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DARB_WITH_ASSERTIONS=ON -DARB_VECTORIZE=ON #-DARB_ARCH=native
make -j
```

I have tried also enabling the ARB_ARCH variable, but again fails. The errors I am getting are of several kinds. Here are some of them:

```
mechanisms/generated/kdrmt_cpu.cpp:118:20: error: no matching constructor for initialization of 'arb::simd_value' (aka 'arb::simd::simd_detail::simd_impl<void>')

arbor/simd/simd.hpp:106:50: error: array has incomplete element type 'const arb::simd::simd_detail::simd_impl<void>::scalar_type' (aka 'const void')

etc.
```
Same behaviour with AppleClang 9.1.0.
Non-vectorized version compiles fine.

--- a/doc/install.rst
+++ b/doc/install.rst
@@ -287,16 +287,17 @@ Vectorization
 -------------
 
 Explicit vectorization of computational kernels can be enabled in Arbor by setting the
-``ARB_VECTORIZE`` CMake flag:
+``ARB_VECTORIZE`` CMake flag. This option is typically used in conjunction with the
+``ARB_ARCH`` option to specify the target architecture: without SIMD support in Arbor
+for the architecture, enabling ``ARB_VECTORIZE`` will lead to a compilation error.
 
 .. code-block:: bash
 
-    cmake -DARB_VECTORIZE=ON
+    cmake -DARB_VECTORIZE=ON -DARB_ARCH=native
 
 With this flag set, the library will use architecture-specific vectorization intrinsics
 to implement these kernels. Arbor currently has vectorization support for x86 architectures
-with AVX, AVX2 or AVX512 ISA extensions. Enabling the `ARB_VECTORIZE` option for a target
-without support in Arbor will give a compilation error.
+with AVX, AVX2 or AVX512 ISA extensions.
 
 .. _gpu:
 

--- a/include/arbor/simd/simd.hpp
+++ b/include/arbor/simd/simd.hpp
@@ -56,6 +56,8 @@ namespace simd_detail {
 
     template <typename Impl>
     struct simd_impl {
+        static_assert(!std::is_void<Impl>::value, "no such SIMD ABI supported");
+
         // Type aliases:
         //
         //     scalar_type           internal value type in one simd lane,

--- a/include/arbor/util/compat.hpp
+++ b/include/arbor/util/compat.hpp
@@ -42,7 +42,7 @@ inline void compiler_barrier_if_icc_leq(unsigned ver) {
 // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85597
 
 template <typename T>
-#if defined(__GNUC__) &&\
+#if !defined(__clang__) && defined(__GNUC__) &&\
     ( __GNUC__<6 ||\
      (__GNUC__==6 && __GNUC_MINOR__*100 + __GNUC_PATCHLEVEL__ < 401) ||\
      (__GNUC__==7 && __GNUC_MINOR__*100 + __GNUC_PATCHLEVEL__ < 301) ||\

